### ISSUE TITLE
.fit getting slower and slower with every epoch

### ISSUE THREAD

**Issue by HWiese1980** [skorch-dev/skorch#568]
Hello,

I'm using `skorch` with a neural network for regression. I'm experiencing a severe slow down that gets worse the more epochs I fit. Partial fit continues this trend. The more epochs have run, the slower the training process gets.

How come?

**Comment by HWiese1980** [comment]
Interesting about this is that it does not seem to be the training itself. The log shows `dur` to be reasonably constant (approximately). Whatever happens between epoch end and the starting of the next epoch is the problem... I'm currently uploading a screencast of the behavior.

**Comment by BenjaminBossan** [comment]
The first peculiar thing is that each epoch takes so much longer then the duration shown in the logs. As there is almost nothing going on between epochs, this strikes me as odd. Since there is no code at all, it's hard to debug for me. E.g., could you tell the batch size, number of samples in train, and number of samples in valid?

Furthermore, could you please test this: 

1) After the net has been initialized, could you remove all callbacks:

```python
net = NeuralNetRegressor(...)
net.initialize()
net.callbacks = []
net.callbacks_ = []
# measure the duration
net.fit(X, y)
```

2) Deactivate the validation and see whether the same slowdown occurs:

```python
net = NeuralNetRegressor(..., train_split=False)
net.fit(X, y)
```

Then measure whether you experience the same slowdown?

Also, I think I've never worked with so many epochs, not even close. Maybe that can somehow be the reason. I wonder if there could be some kind of inefficiency in, say, the `history` implementation that makes it slow with so many entries.

**Comment by HWiese1980** [comment]
Thanks. 

I've overridden the print_log callback in order to not show every single epoch outcome. It only shows every 100th output. The time between the outputs is the time the other 99 epochs in between are taking. The actual issue is that this time is drastically increasing after some 100 epochs. 

I'll give your code a shot and share some code beginning next week when I'm back at work.

**Comment by HWiese1980** [comment]
Sorry I haven't gotten back to you yet. My priorities have shifted since my previous message. I hope I'll be able to send you the code soon.

**Comment by BenjaminBossan** [comment]
Don't worry, we appreciate any help.

**Comment by BenjaminBossan** [comment]
@HWiese1980 Did you have time to dig deeper yet?

**Comment by HWiese1980** [comment]
@BenjaminBossan not yet, unfortunately. I'm sorry. I'll try and see what I can do. Best to keep the issue open for me to be reminded every now and then.

**Comment by BenjaminBossan** [comment]
@HWiese1980 reminder! :)

**Comment by HWiese1980** [comment]
Thank you. I hope that I'll be able to get back to it on next week's Monday...

**Comment by BenjaminBossan** [comment]
@HWiese1980 Did you have a time to have another look yet?

**Comment by HWiese1980** [comment]
Not yet, I'm terribly sorry that it takes so long. I'll soon be back on the topic and hope to be able to reproduce the issue.

**Comment by jjxiao-fn** [comment]
I seem to see similar issues with the fit method. The fit method seems about 7~10x slower than standard PyTorch training cycles calling the model module. Wondering if there is a way to not say the history for every epoch..

**Comment by BenjaminBossan** [comment]
@jjxiao-fn Thanks for reporting. Could you provide a bit of context? Ideally, you could give some code that would allow us to reproduce the result.

If that's not possible, maybe you can answer these questions:

1. what exactly did you measure
2. is the module very "small" compared to the overhead (e.g. cycling through the data)
3. do you train a large number of epochs
4. do you use cuda? which GPU?
5. could you try turning off all callbacks?

```python
net = NeuralNet(...)
net.initialize()
net.callbacks_ = []
net.fit(X, y)
```

We have a [benchmark based on mnist](https://github.com/skorch-dev/skorch/blob/master/examples/benchmarks/mnist.py). I just ran it and there is essentially no difference between pure pytorch and skorch. Therefore, if you see a difference, there must be something quite different going on.

**Comment by jjxiao-fn** [comment]
@BenjaminBossan Thank you for your quick reply. Here is an example showing you what I was seeing and how I measured.
[skorch.pdf](https://github.com/skorch-dev/skorch/files/4467373/skorch.pdf)
 From the notebook, you can see skorch seemed to take 1.23s to fit the data with 150 epochs, while the torch version just took 96.2ms to iterate the same number of epochs. You can also see the performance is much higher if cross-entropy loss was used. Wondering that was related to https://github.com/skorch-dev/skorch/issues/519 and https://discuss.pytorch.org/t/trouble-getting-probability-from-softmax/26764/6. 
Executable codes are available in the zip file. 
[skorch_code.zip](https://github.com/skorch-dev/skorch/files/4467383/skorch_code.zip)

BTW, I didn't use cuda yet and I have tried to turn off callbacks but they didn't seem to speedup skorch in my tests.

**Comment by BenjaminBossan** [comment]
@jjxiao-fn Thank you for reporting back.

From a look at your notebook, this is my conclusion: You use a very small dataset with a very small module. In this setting, the skorch overhead will be relatively big. The reason is that skorch will add a fixed-size overhead (and a very minor variable-sized overhead, but let's ignore this for now). The fixed overhead may just be a second or so, but when the overall training just takes 1.2 seconds, that's a big chunk. I'm pretty sure that if you use a bigger dataset or a more complex module, the skorch overhead should become negligible.

Regarding the issue with softmax being unstable: You can also use logsoftmax + crossentropy loss with skorch, that's no problem. Again, however, I'm sure that with more data and a more reasonable architecture, the issue will probably vanish.

**Comment by BenjaminBossan** [comment]
I will close this for now, assuming the problem has been resolved, as there is no further feedback.
